X11R6/contrib/programs/xdaliclock-2.03/xdaliclock.man
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH XDaliClock 1 "15-May-94" "X Version 11"
.SH NAME
xdaliclock - melting digital clock
.SH SYNOPSIS
.B xdaliclock
[-\fItoolkitoption\fP ...] [-\fIoption\fP ...]
.SH DESCRIPTION
The \fIxdaliclock\fP program displays a digital clock; when a digit
changes, it ``melts'' into its new shape.

This program was inspired by the Alto and Macintosh programs of the same
name, written by Steve Capps in 1983 or 1984.
.SH OPTIONS
.I xdaliclock
accepts all of the standard toolkit options, and also accepts the
following options:
.TP 8
.B \-help
Print a brief summary of the allowed options on the standard error output.
.TP 8
.B \-12
Use a twelve hour clock.
.TP 8
.B \-24
Use a twenty-four hour clock.
.TP 8
.B \-seconds
Update every second.
.TP 8
.B \-noseconds
Update once per minute; don't display seconds at all.
.TP 8
.B \-cycle
Do color-cycling.
.TP 8
.B \-nocycle
Don't do color-cycling.
.TP 8
.B \-visual \fIvisual\fP
Specify which visual to use.  Legal values are:
.RS 8
.TP 8
.B best
Use the visual which supports the most writable color cells; this is
the default.
.TP 8
.B default
Use the screen's default visual (the visual of the root window.)  This is
not necessarily the most colorful visual, which is why it is not the default.
.TP 8
.I class
One of \fBStaticGray\fP, \fBStaticColor\fP, \fBTrueColor\fP, \fBGrayScale\fP, 
\fBPseudoColor\fP, or \fBDirectColor\fP.  Selects the deepest visual of
the given class.
.TP 8
.I number
A number (decimal or hex) is interpreted as a visual id number, as reported
by the
.BR xdpyinfo (1)
program; in this way you can select a shallower visual if desired.
.RE
.TP 8
.B \-shape
Use the Shape Extension, if it's available, to make the window background be
transparent.  

This doesn't interact terribly well with the \fItwm\fP or \fItvtwm\fP window
managers, unless they have been configured to not put a titlebar on
the \fIxdaliclock\fP window.  If this isn't specified in your \fI.twmrc\fP
or \fI.tvtwmrc\fP file, then the window will flicker constantly, as the
window manager tries to add and remove the titlebar ten times each second.

Also, this turns most X servers into huge cycle hogs.  This is probably
because of inefficient implementations of the Shape extension.
.TP 8
.B \-noshape
Don't use the Shape Extension.
.TP 8
.B \-memory low
Use high-bandwidth, low-memory mode.  If you have a very fast connection
between the machine this program is running on and the X server it is
displaying on, then \fIxdaliclock\fP can work correctly by simply making
the drawing requests it needs when it needs them.  This is \fIthe elegant
method.\fP  However, the amount of data necessary to animate the display 
ends up being a bit over 10 kilobytes worth of X Protocol per second.  On a
fast machine with a local display, or over a fast network, that's almost
negligible, but (for example) an NCD X Terminal at 38.4 kilobaud can't 
keep up.  That is the reason for:
.TP 8
.B \-memory medium
Use high-memory, low-bandwidth mode.  In this mode, \fIxdaliclock\fP
precomputes most of the frames that it will ever need.  This is \fIthe
sleazy copout method.\fP  The bandwidth requirements are drastically 
reduced, because instead of telling the server what bits to draw where, 
it merely tells it what pixmaps to copy into the window.  Aside from the 
fact that I consider this to be \fIcheating,\fP the only downside of this
method is that those pixmaps (about 170 of them, each the size of one
character) are consuming server-memory.  This probably isn't a very big deal,
unless you're using an exceptionally large font.
.TP 8
.B \-memory high
With \fImemory\fP set to \fIhigh\fP, the cache is twice as 
large (the \fIn -> n+2\fP transitions are cached as well as the \fIn -> n+1\fP
ones).  Even with memory set to \fImedium\fP, this program can seem sluggish 
when using the builtin font over a very slow connection to the display server.
.TP 8
.B \-font \fIfontname\fP
Specifies the X font to use; \fIxdaliclock\fP can correctly animate any font
that contains all the digits plus colon and slash, and in which the letters
aren't excessively curly.

The \fIxdaliclock\fP program contains a pair of builtin bitmapped fonts,
which are larger and more attractive than the standard X fonts.  One of
these fonts will be used if the \fI-font\fP option is given one of the
fontnames \fBBUILTIN\fP or \fBBUILTIN2\fP.
.TP 8
.B \-builtin
This is the same as specifying \fI-font BUILTIN\fP.
.TP 8
.B \-builtin2
This is the same as specifying \fI-font BUILTIN2\fP.
.TP 8
.B \-fullscreen
Make the window take up the whole screen.  When \fI-fullscreen\fP is 
specified, the displayed time will wander around a little, to prevent
any pixels from being on continuously and causing phosphor burn-in.
.TP 8
.B \-root
Display the clock on the root window instead of in its own window.
This makes the digits wander around too.
.PP
The following standard X Toolkit command line arguments are commonly 
used with \fIxdaliclock\fP:
.TP 8
.B \-display \fIhost:dpy\fP
This option specifies the X server to contact.
.TP 8
.B \-geometry \fIgeometry\fP
This option specifies the prefered size and position of the clock window.
.TP 8
.B \-bg \fIcolor\fP
This option specifies the color to use for the background of the window.
The default is ``white.''
.TP 8
.B \-fg \fIcolor\fP
This option specifies the color to use for the foreground of the window.
The default is ``black.''
.TP 8
.B \-bd \fIcolor\fP
This option specifies the color to use for the border of the window.
The default is the same as the foreground color.
.TP 8
.B \-rv
This option indicates that reverse video should be simulated by swapping
the foreground and background colors.
.TP 8
.B \-bw \fInumber\fP
This option specifies the width in pixels of the border surrounding 
the window.
.TP 8
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.
.SH COMMANDS
Clicking and holding any mouse button in the \fIxdaliclock\fP window will
cause it to display the date while the button is held.
.PP
Typing ``space'' at the \fIxdaliclock\fP window will toggle between a 
twelve hour and twenty-four hour display.
.PP
Typing ``q'' or ``C-c'' at the window quits.

If the \fIxdaliclock\fP window is iconified or otherwise unmapped, it
will go to sleep until it is mapped again.
.SH X DEFAULTS
\fIxdaliclock\fP understands all of the core resource names and
classes as well as:
.PP
.TP 8
.B seconds (\fPclass\fB Seconds)
Whether to display seconds.  
If true, this is the same as the \fI-seconds\fP command line argument;
if false, this is the same as \fI-noseconds.\fP
.TP 8
.B cycle (\fPclass\fB Cycle)
Whether to do color cycling.
If true, this is the same as the \fI-cycle\fP command line argument;
if false, this is the same as \fI-nocycle.\fP
.TP 8
.B shape (\fPclass\fB Shape)
Whether to use the Shape Extension, if available.
If true, this is the same as the \fI-shape\fP command line argument;
if false, this is the same as \fI-noshape.\fP
.TP 8
.B memory (\fPclass\fB Memory)
This must be \fIhigh\fP, \fImedium\fP, or \fIlow\fP, the same as
the \fI-memory\fP command-line option.
.TP 8
.B font (\fPclass\fB Font)
The same as the \fI-font\fP command line option: the font to melt.
If this is the string \fBBUILTIN\fP, then the large builtin font is
used.  If this is the string \fBBUILTIN2\fP, then the \fIeven larger\fP
builtin font is used.  Otherwise, this must be the name of a valid X font.
.TP 8
.B mode (\fPclass\fB Mode)
Whether to display 12-hour or 24-hour time.
If \fB12\fP, this is the same as the \fI\-12\fP command line argument;
if \fB24\fP, this is the same as \fI\-24.\fP
.TP 8
.B datemode (\fPclass\fB DateMode)
Specifies how the date should be printed when a mouse button is held down.
This may be one of the strings \fBmm/dd/yy\fP, \fBdd/mm/yy\fP,
\fByy/mm/dd\fP, \fByy/dd/mm\fP, \fBmm/yy/dd\fP, or \fBdd/yy/mm\fP.
The default is \fBmm/dd/yy\fP.  If seconds are not being displayed, then
only the first four digits will ever be displayed (\fBmm/dd\fP instead
of \fBmm/dd/yy\fP, for example.)
.TP 8
.B fullScreen (\fPclass\fB FullScreen)
The same as the \fI\-fullscreen\fP command-line option.
.TP 8
.B root (\fPclass\fB Root)
The same as the \fI\-root\fP command-line option.
.TP 8
.B visualID (\fPclass\fB VisualID)
The same as the \fI\-visual\fP command-line option.
.SH ENVIRONMENT
.PP
.TP 4
.B DISPLAY
to get the default host and display number.
.TP 4
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 4
.B TZ
to get the current time zone.  If you want to force the clock to display
some other time zone, set this variable before starting it.  For example:
.EX
$ TZ=GMT0 xdaliclock
% ( setenv TZ PST8PDT ; xdaliclock )
.EE
You may notice that the format of the TZ variable (which is used by the
C library
.BR ctime (3)
and
.BR localtime (3)
routines) is not actually documented anywhere.  The fourth character (the
digit) is the only thing that really matters: it is the offset in hours from
GMT.  The first three characters are ignored.  The last three characters are
used to flag daylight savings time: their presence effectively adds 1 to the
zone offset.  (I am not making this up...)
.SH "SEE ALSO"
.BR X (1),
.BR xrdb (1),
.BR xlsfonts (1),
.BR xclock (1),
.BR dclock (1),
.BR oclock (1),
.BR tclock (1),
.BR xscreensaver (1)
.SH BUGS
Other system load will sometimes cause the second-display to increment
by more than one second at a time, in order to remain synchronized to 
the current time.

The \fI-memory\fP option is disgusting and shouldn't be necessary, but I'm
not clever enough to eliminate it.  It has been said that hacking graphics
in X is like finding \fIsqrt(pi)\fP with roman numerals.

When using a small font (less than 48x56 or so) it's possible that shipping
a bitmap to the server would be more efficient than sending a DrawSegments
request (since the endpoints are specified using 16 bits each, when all that
we really need is 6 or 7 bits.)

Support for the Shared Memory Extension would be a good thing.

It should display the day of the week somewhere.

The color cycling should be less predictable; it should vary saturation and
intensity as well, and should be more careful that foreground and background
contrast well.

Should have a \fI-analog\fP mode (maybe someday...)
.SH COPYRIGHT
Copyright \(co 1991, 1992, 1993, 1994 by Jamie Zawinski.  Permission to use, 
copy, modify, distribute, and sell this software and its documentation for 
any purpose is hereby granted without fee, provided that the above copyright
notice appear in all copies and that both that copyright notice and this
permission notice appear in supporting documentation.  No representations are
made about the suitability of this software for any purpose.  It is provided
"as is" without express or implied warranty.
.SH AUTHOR
Jamie Zawinski <jwz@lucid.com>, 18-sep-91.

Please let me know if you find any bugs or make any improvements.

Thanks to Ephraim Vishniac <ephraim@think.com> for explaining the format of
the bitmap resources in the Macintosh version of this, so that I could snarf
them for the \fI-builtin\fP fonts.

And thanks to Steve Capps for the really great idea.


X11R6/contrib/programs/xdaliclock-2.03/xdaliclock.c
/* xdaliclock - a melting digital clock
 * Copyright (c) 1991, 1992, 1993, 1994 Jamie Zawinski <jwz@lucid.com>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or
 * implied warranty.
 *
 * Version 1 of this program used only Xlib, not Xt.  The dynamically linked
 * sparc executable was 81k, and geometry and resource handling were kludgy
 * and broken, because I had to duplicate the obscure, undocumented things
 * that Xt does.  Version 2 uses Xt, and geometry and resource handling work
 * correctly - but the dynamically linked executable has more than tripled in
 * size, swelling to 270k.  This is what is commonly refered to as "progress."
 * (Bear in mind that the first (or second) program ever to implement this
 * algorithm ran perfectly well on a Macintosh with a 128k address space.)
 */

#include "version.h"

#include <stdio.h>
#include <X11/Intrinsic.h>
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include <X11/CoreP.h>

#include "xdaliclock.h"
#include "defaults.h"

char *progname;
char *progclass;
char *hacked_version;
XrmDatabase db;
Bool wander_p;
Bool root_p;
Bool do_cycle;

static void usage P((void));
static void hack_version P((void));

static XrmOptionDescRec options [] = {
  /* global options */
  { "-root",		".root",		XrmoptionNoArg, "True" },
  { "-onroot",		".root",		XrmoptionNoArg, "True" },
  { "-window",		".root",		XrmoptionNoArg, "False" },
  { "-fullscreen",	".fullScreen",		XrmoptionNoArg, "True" },
  { "-mono",		".mono",		XrmoptionNoArg, "True" },
  { "-visual",		".visualID",		XrmoptionSepArg, 0 },
  { "-install",		".installColormap",	XrmoptionNoArg, "True" },
  { "-title",		".title",		XrmoptionSepArg, 0  }, /* R3 */
  { "-xrm",		0,			XrmoptionResArg, 0 },

  /* options from digital.c */
  { "-12",		".mode",		XrmoptionNoArg, "12" },
  { "-24",		".mode",		XrmoptionNoArg, "24" },
  { "-datemode",	".datemode",		XrmoptionSepArg, 0 },
  { "-mm/dd/yy",	".datemode",		XrmoptionNoArg, "MM/DD/YY" },
  { "-dd/mm/yy",	".datemode",		XrmoptionNoArg, "DD/MM/YY" },
  { "-seconds",		".seconds",		XrmoptionNoArg, "True" },
  { "-noseconds",	".seconds",		XrmoptionNoArg, "False" },
  { "-cycle",		".cycle",		XrmoptionNoArg, "True" },
  { "-nocycle",		".cycle",		XrmoptionNoArg, "False" },
  { "-font",		".font",		XrmoptionSepArg, 0 },
  { "-fn",		".font",		XrmoptionSepArg, 0 },
  { "-builtin",		".font",		XrmoptionNoArg, "BUILTIN" },
  { "-builtin2",	".font",		XrmoptionNoArg, "BUILTIN2" },
  { "-memory",		".memory",		XrmoptionSepArg, 0 },
  { "-sleaze-level",	".memory",		XrmoptionSepArg, 0 },
  { "-oink",		".memory",		XrmoptionNoArg, "Medium" },
  { "-oink-oink",	".memory",		XrmoptionNoArg, "High" },
  { "-shape",		".shape",		XrmoptionNoArg, "True" },
  { "-noshape",		".shape",		XrmoptionNoArg, "False" }
};

int
main (argc, argv)
     int argc;
     char **argv;
{
  Widget toplevel;
  Display *dpy;
  unsigned int w, h;
  Window window;
  Visual *visual;
  Colormap cmap;
  unsigned long fg, bg, bd;
  XWindowAttributes xgwa;
  Bool full_screen_p;
  XEvent event;

  hack_version ();
  progclass = "XDaliClock";

#if (XtSpecificationRelease >= 4)
  {
    XtAppContext app;
    toplevel = XtAppInitialize (&app, progclass, options,
				sizeof (options) / sizeof (options [0]),
				&argc, argv, defaults, 0, 0);
    dpy = XtDisplay (toplevel);
    db = XtDatabase (dpy);
    XtGetApplicationNameAndClass (dpy, &progname, &progclass);
  }
#else /* R3 */
  {
    char *tmp;
    int i, j;
    XrmDatabase argv_db, server_db, fallback_db;
    progname = argv [0];
# ifdef VMS
    while (tmp = (char *) strrchr (progname, ']'))
      progname = tmp + 1;
    if (tmp = (char *) strrchr (progname, '.'))
      *tmp = '\0';
# else /* !VMS */
    if (tmp = (char *) strrchr (progname, '/'))
      progname = tmp + 1;
# endif /* !VMS */

    argv_db = 0;
    XrmParseCommand (&argv_db,
		     options, sizeof (options) / sizeof (options [0]),
		     progname, &argc, argv);
    toplevel = XtInitialize (progname, progclass, options,
			     sizeof (options) / sizeof (options [0]),
			     &argc, argv);
    dpy = XtDisplay (toplevel);
    server_db = XtDatabase (dpy);
    for (i = 0, j = 0; defaults [i]; i++)
      j += strlen (defaults [i]) + 1;
    tmp = (char *) malloc (j + 1);
    for (i = 0; defaults [i]; i++)
      {
	tmp = strcat (tmp, defaults [i]);
	tmp += strlen (tmp);
	*tmp++ = '\n';
	*tmp = 0;
      }
    tmp -= j;
    fallback_db = XrmGetStringDatabase (tmp);
    free (tmp);
    db = fallback_db;
    XrmMergeDatabases (server_db, &db);
    XrmMergeDatabases (argv_db, &db);
  }
#endif /* !R3 */

  if (argc > 1)
    {
      if (strcmp (argv [1], "-help"))
	fprintf (stderr, "%s: unrecognised option \"%s\"\n",
		 progname, argv[1]);
      usage ();
      exit (-1);
    }

  root_p = get_boolean_resource ("root", "Boolean");
  full_screen_p = get_boolean_resource ("fullScreen", "Boolean");
  do_cycle = get_boolean_resource ("cycle", "Cycle");

  if (root_p)
    {
      visual = DefaultVisual (dpy, DefaultScreen (dpy));
      cmap = DefaultColormap (dpy, DefaultScreen (dpy));
    }
  else
    {
      /* Note: we can't use the resource ".visual" because Xt is SO FUCKED. */
      visual = get_visual_resource (dpy, "visualID", "VisualID");
      if ((visual != DefaultVisual (dpy, DefaultScreen (dpy))) ||
	  get_boolean_resource ("installColormap", "InstallColormap"))
	cmap = XCreateColormap (dpy, RootWindow (dpy, DefaultScreen (dpy)),
				visual, AllocNone);
      else
	cmap = DefaultColormap (dpy, DefaultScreen (dpy));
    }

  wander_p = (root_p || full_screen_p);

  initialize_digital (dpy, visual, cmap, &fg, &bg, &bd, &w, &h);

  if (root_p)
    {
      window = RootWindowOfScreen (XtScreen (toplevel));
      XtDestroyWidget (toplevel);
      XGetWindowAttributes (dpy, window, &xgwa);
      cmap = xgwa.colormap;
      visual = xgwa.visual;
    }
  else
    {
      int xmargin = 10;
      int ymargin = 10;
      Dimension ow, oh, bw;
      Arg av [20];
      int ac;

      ac = 0;
      XtSetArg (av [ac], XtNinput, True); ac++;
      XtSetArg (av [ac], XtNvisual, visual); ac++;
      XtSetArg (av [ac], XtNcolormap, cmap); ac++;
      XtSetArg (av [ac], XtNforeground, ((Pixel) fg)); ac++;
      XtSetArg (av [ac], XtNbackground, ((Pixel) bg)); ac++;
      XtSetArg (av [ac], XtNborderColor, (Pixel) bd); ac++;
      /* This sometimes causes -geometry to be ignored... */
      /* XtSetArg (av [ac], XtNminWidth, (Dimension) w); ac++; */
      /* XtSetArg (av [ac], XtNminHeight, (Dimension) h); ac++; */
      XtSetValues (toplevel, av, ac);

      ac = 0;
      XtSetArg (av [ac], XtNwidth, &ow); ac++;
      XtSetArg (av [ac], XtNheight, &oh); ac++;
      XtSetArg (av [ac], XtNborderWidth, &bw); ac++;
      XtGetValues (toplevel, av, ac);

      if (full_screen_p)
	{
	  w = DisplayWidth  (dpy, DefaultScreen (dpy));
	  h = DisplayHeight (dpy, DefaultScreen (dpy));
	}
      if (ow <= 1 || oh <= 1)
	{
	  Dimension ww = w + xmargin + xmargin;
	  Dimension hh = h + ymargin + ymargin;
	  ac = 0;
	  XtSetArg (av [ac], XtNwidth,  ww); ac++;
	  XtSetArg (av [ac], XtNheight, hh); ac++;
	  XtSetValues (toplevel, av, ac);
	}
      if (full_screen_p)
	{
	  /* #### Want this to set UPosition, not PPosition... */
	  ac = 0;
	  XtSetArg (av [ac], XtNx, -xmargin); ac++;
	  XtSetArg (av [ac], XtNy, -ymargin); ac++;
	  XtSetValues (toplevel, av, ac);
	}

      XtRealizeWidget (toplevel);
      window = XtWindow (toplevel);
    }

  run_digital (dpy, window);
  return (0);
}


static void
usage ()
{
  fprintf (stderr, "%s\n\
usage: %s [ options ]\n\
where options are\n\
\n\
  -12				Display twelve hour time (default).\n\
  -24				Display twenty-four hour time.\n\
  -seconds			Display seconds (default).\n\
  -noseconds			Don't display seconds.\n\
  -cycle			Do color-cycling.\n\
  -nocycle			Don't do color-cycling (default).\n\
  -display <host:dpy>		The display to run on.\n\
  -visual <visual-class>	The visual to use.\n\
  -geometry <geometry>		Size and position of window.\n\
  -foreground or -fg <color>	Foreground color (default black).\n\
  -background or -bg <color>	Background color (default white).\n\
  -reverse or -rv		Swap foreground and background.\n\
  -borderwidth or -bw <int>	Border width.\n\
  -bd or -border <color>	Border color.\n\
  -title <string>		Window title.\n\
  -name <string>		Resource-manager class name (XDaliClock).\n\
  -fullscreen			Use a window that takes up the whole screen.\n\
  -root				Draw on the root window instead.\n\
", version + 4, progname);
#if defined (BUILTIN_FONT) && defined (BUILTIN_FONT_2)
  fprintf (stderr, "\
  -font	or -fn <font>		Name of an X font to use, or the string\n\
				\"BUILTIN\", meaning to use the large builtin\n\
				font (this is the default), or \"BUILTIN2\",\n\
				meaning to use the even larger builtin font.\n\
  -builtin			Same as -font BUILTIN.\n\
  -builtin2			Same as -font BUILTIN2.\n\
");
#else /* ! (BUILTIN_FONT && BUILTIN_FONT_2) */
#if (defined (BUILTIN_FONT) || defined (BUILTIN_FONT_2))
  fprintf (stderr, "\
  -font	or -fn <font>		Name of an X font to use, or the string\n\
				\"BUILTIN\", meaning to use the large builtin\n\
				font (this is the default).\n\
  -builtin			Same as -font BUILTIN.\n\
");
#else /* ! (BUILTIN_FONT || BUILTIN_FONT_2) */
  fprintf (stderr, "\
  -font	or -fn <font>		Name of an X font to use.\n\
");
#endif /* ! (BUILTIN_FONT || BUILTIN_FONT_2) */
#endif /* ! (BUILTIN_FONT && BUILTIN_FONT_2) */
  fprintf (stderr, "\
  -memory <high|medium|low>	Tune the memory versus bandwidth consumption;\n\
				default is \"low\".\n\
  -oink				Same as -memory medium.\n\
  -oink-oink			Same as -memory high.\n\
");
#ifdef SHAPE
  fprintf (stderr, "\
  -shape			Use the Shape extension if available.\n\
  -noshape			Don't.\n");
#endif
#if !(defined (BUILTIN_FONT) || defined (BUILTIN_FONT_2)) || !defined (SHAPE)
  fprintf (stderr, "\n");
# ifndef BUILTIN_FONT
   fprintf(stderr, "This version has been compiled without the builtin fonts");
# endif
# ifndef SHAPE
#  ifndef BUILTIN_FONT
  fprintf (stderr, ", or support for the\nShape Extension.\n");
#  else
  fprintf (stderr, "\
This version has been compiled without support for the Shape Extension.\n");
#  endif
# else
  fprintf (stderr, ".\n");
# endif
#endif
  exit (-1);
}

static void
hack_version ()
{
  char *src = version + 4;
  char *dst = hacked_version = (char *) malloc (strlen (src) + 1);
#define digitp(x) (('0' <= (x)) && ((x) <= '9'))
  while (*src)
    if (!strncmp ("Copyright (c)", src, 13))
      *dst++ = '\251', src += 13;
    else if (*src == '(') *dst++ = '<', *src++;
    else if (*src == ')') *dst++ = '>', *src++;
    else if (digitp (src[0]) && digitp (src[1]) && digitp (src[2]) &&
	     digitp (src[3]) && src[4] == ',' && src[5] == ' ')
      src += 6;
    else
      *dst++ = *src++;
  *dst = 0;
}

#ifdef VMS
/* this isn't right, but it's good enough (only returns 1 or 0) */
int
strcasecmp (s1, s2)
    char *s1, *s2;
{
  int i;
  int L1 = strlen (s1);
  int L2 = strlen (s2);
  if (L1 != L2)
    return 1;
  for (i = 0; i < L1; i++)
    if ((isupper (s1[i]) ? _tolower (s1[i]) : s1[i]) !=
	(isupper (s2[i]) ? _tolower (s2[i]) : s2[i]))
      return 1;
  return 0;
}
#endif /* VMS */
